src/m2n/PointToPointCommunication.cpp

```diff
@@ -245,42 +281,65 @@ void printLocalIndexCountStats(std::map<int, std::vector<int>> const &m)
+ * the complexity boils down to linear complexity
+ * \f$ \mathcal{O}(p n) \f$
  */
 std::map<int, std::vector<int>> buildCommunicationMap(
     // `thisVertexDistribution' is input vertex distribution from this participant.
-    mesh::Mesh::VertexDistribution const &thisVertexDistribution,
+    mesh::Mesh::VertexDistribution &thisVertexDistribution,
     // `otherVertexDistribution' is input vertex distribution from other participant.
-    mesh::Mesh::VertexDistribution const &otherVertexDistribution,
-    int                                   thisRank = utils::IntraComm::getRank())
+    mesh::Mesh::VertexDistribution &otherVertexDistribution,
+    int                             thisRank = utils::IntraComm::getRank())
 {
   auto iterator = thisVertexDistribution.find(thisRank);
   if (iterator == thisVertexDistribution.end())
     return {};
 
-  // Build lookup table from otherIndex -> rank for the otherVertexDistribution
-  const auto lookupIndexRank = [&otherVertexDistribution] {
-    boost::container::flat_multimap<int, int> lookupIndexRank;
-    for (const auto &other : otherVertexDistribution) {
-      for (const auto &otherIndex : other.second) {
-        lookupIndexRank.emplace(otherIndex, other.first);
-      }
-    }
-    return lookupIndexRank;
-  }();
+  std::map<int, std::vector<int>> communicationMap;
+
+  // take advantage that these data structures are in most cases sorted by construction,
+  // i.e., we perform here mostly a safety check and don't perform an actual sorting
+  if (!std::is_sorted(iterator->second.begin(), iterator->second.end())) {
+    std::sort(iterator->second.begin(), iterator->second.end());
+  }
 
-  auto const &indices = iterator->second;
+  // now we iterate over all other vertex distributions to compute the intersection
+  for (auto &other : otherVertexDistribution) {
+    // first a safety check, that we are actually sorted, similar to above
+    if (!std::is_sorted(other.second.begin(), other.second.end())) {
+      std::sort(other.second.begin(), other.second.end());
+    }
 
-  std::map<int, std::vector<int>> communicationMap;
-  for (size_t index = 0lu; index < indices.size(); ++index) {
-    auto range = lookupIndexRank.equal_range(indices[index]);
-    for (auto iter = range.first; iter != range.second; ++iter) {
-      communicationMap[iter->second].push_back(index);
+    // before starting to compute an actual intersection, we first check if elements can
+    // possibly be in both data sets by comparing upper and lower index bounds of both
+    // data sets. For typical partitioning schemes, each rank only exchanges data with
+    // a few neighbors such that this check already filters out a significant amount of
+    // computations
+    if (iterator->second.empty() || other.second.empty() || (other.second.back() < iterator->second.at(0)) || (other.second.at(0) > iterator->second.back())) {
+      // in this case there is nothing to be done
+      continue;
+    } else {
```

Comment:
This else isn't necessary, as `continue` already skips the rest here. Reducing indentation is always a good thing :wink:

Reply:
Yes, belongs to the series of 'being to deep into coding' :sweat_smile: